redbot/core/utils/__init__.py

```diff
@@ -255,3 +256,54 @@ def bounded_gather(
     tasks = (_sem_wrapper(semaphore, task) for task in coros_or_futures)
 
     return asyncio.gather(*tasks, return_exceptions=return_exceptions)
+
+
+class AsyncGen(AsyncIterable[_T]):
```

Comment:
This class is not just `AsyncIterable` but also `AsyncIterator` so that generic suits it better. It can also be awaited so I think you should add `Awaitable` generic as well.
```suggestion
class AsyncGen(AsyncIterator[_T], Awaitable[List[_T]]):  # pylint: disable=duplicate-bases
```